File: dbm_conv.hs
-- dBm to volt and volt to dBm converter
-- 50 Ohm load

module Main
  where

import IO
import Text.Printf

main = do
  putStrLn "dBm convertor v1.0.0 ; 50 Ohm load"
  hSetBuffering stdin LineBuffering
  getChoice

getChoice = do
  printChoices
  num <- getLine
  let choiceNum = read num
  case choiceNum of
    0 -> return ()
    1 -> do putStrLn "Vpk-pk (V) ? : "
            result <- getValue 1
            putStrLn $ printf "\n  %f V pk-pk => %.4f dBm" (fst result:: Double) (snd result)
            getChoice
    2 -> do putStrLn "V rms (V) ? : "
            result <- getValue 2
            putStrLn $ printf "\n  %f V rms => %.4f dBm" (fst result:: Double) (snd result)
            getChoice
    3 -> do putStrLn "dbm ? : "
            result <- getValue 3
            putStrLn $ printf "\n  %f dBm => %.4f V pk-pk" (fst result:: Double) (snd result)
            getChoice
    4 -> do putStrLn "dbm ? : "
            result <- getValue 4
            putStrLn $ printf "\n  %f dBm => %.4f V pk-pk" (fst result:: Double) (snd result)
            getChoice
    _ -> getChoice

printChoices = do
  putStrLn " "
  putStrLn "select : "
  putStrLn "  0 : quit"
  putStrLn "  1 : V pk-pk to dBm"
  putStrLn "  2 : V rms to dBm"
  putStrLn "  3 : dBm to V pk-pk"
  putStrLn "  4 : dBm to V rms"

getValue convertionChoice = do
  numValue <- getLine
  let value = read numValue
  converter convertionChoice value

converter convertionChoice value
  | convertionChoice == 1 = return ((value , 10.0 * logBase 10 (value^2 * 2.5)))
  | convertionChoice == 2 = return ((value , 10.0 * logBase 10 (value^2 * 20)))
  | convertionChoice == 3 = return ((value , sqrt (0.4 * 10**(value / 10.0))))
  | convertionChoice == 4 = return ((value , sqrt (0.05 * 10**(value / 10.0))))
  | otherwise             = return (0.0, 0.0)

File: dBmConv.py
# dBm to volt and volt to dBm converter
# 50 Ohm load
# License: LGPLv3

from tkinter import *
from math    import log10, sqrt

################################################################################
# Get the input value
################################################################################
def convert (event) :
  try :
    val = eval(entryField.get())
    if convChoice.get() == 1 :
      convertedValue = "{:.3f} V pk-pk <=> {:.3f} dBm"
      result.configure(text = convertedValue.format(val, 10.0 * log10(val**2 * 2.5)))
    elif convChoice.get() == 2 :
      convertedValue = "{:.3f} V rms <=> {:.3f} dBm"
      result.configure(text = convertedValue.format(val, 10.0 * log10(val**2 * 20)))
    elif convChoice.get() == 3 :
      convertedValue = "{:.3f} dBm <=> {:.3f} V pk-pk"
      result.configure(text = convertedValue.format(val, sqrt(0.4 * 10**(val / 10.0))))
    elif convChoice.get() == 4 :
      convertedValue = "{:.3f} dBm <=> {:.3f} V rms"
      result.configure(text = convertedValue.format(val, sqrt(0.05 * 10**(val / 10.0))))
    else :
      result.configure(text = 'Unexpected convertion, error in soft !')
  except :
    result.configure(text = 'Bad value, try a number or a voltage differant of 0V!')

################################################################################
################################  Main  ########################################
################################################################################
convWindows = Tk()
convWindows.title('Volt <=> dBm ; v1.0.0')

# Title and version
Label(convWindows, text = 'Volt to dBm convertor on a 50 Ohm load').grid(row  = 0, column = 0, columnspan = 2, padx = 10)

# Radio buttons
convChoice = IntVar()
R1 = Radiobutton(convWindows, text = "V pk-pk to dBm", variable = convChoice, value=1)
R1.grid(row = 2, column = 0, sticky = W, padx = 10 , pady = 5)
R1.select()
R2 = Radiobutton(convWindows, text = "V rms to dBm", variable = convChoice, value=2)
R2.grid(row = 3, column = 0, sticky = W, padx = 10)
R3 = Radiobutton(convWindows, text = "dBm to V pk-pk", variable = convChoice, value=3)
R3.grid(row = 2, column = 1, sticky = W, padx = 10 , pady = 5)
R4 = Radiobutton(convWindows, text = "dBm to V rms", variable = convChoice, value=4)
R4.grid(row = 3, column = 1, sticky = W, padx = 10)

# Value entry
entryField = Entry(convWindows)
entryField.bind("<Return>", convert)
entryField.grid(row = 4, column = 0, columnspan = 2, padx = 10 , pady = 5)

# Print result
result = Label(convWindows, text = 'xxx <=> yyy', relief = GROOVE, bg = '#F0F0E0', bd = 2)
result.grid(row = 5, column = 0, columnspan = 2, padx = 10, pady = 5)

# Perform calculation button
Button(convWindows, text = 'Convert', command = lambda : convert('0')).grid(row = 6, column = 0, pady = 5)

# Quit button
Button(convWindows, text='Quit', command = convWindows.destroy).grid(row = 6, column = 1, pady = 5)

convWindows.mainloop()
